[WebDriver][socket] Titles containing multibyte characters cannot be retrieved correctly

haruhisa-shin opened the pull request:
#### 0a3175f4f36add1cbc3e595fa45ce59081303f8c
<pre>
[WebDriver][socket] Titles containing multibyte characters cannot be retrieved correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=270063">https://bugs.webkit.org/show_bug.cgi?id=270063</a>

Reviewed by Alexey Proskuryakov.

If the document title contains multibyte characters such as Japanese
or entity references, the &quot;Get Title&quot; result will be garbled.

The title string is obtained by JavaScript&apos;s &quot;document.title()&quot;,
and this data is encoded in UTF8.
However, the StringBuilder.append() function used to create HTTP messages
uses fromLatin1() internally to generate strings from byte data.
This seems to be causing the multibyte characters to be garbled.

This patch changes to use String::fromUTF8() before concatenation to
restore the correct WTF::String even if it contains multibyte characters.

Also, the change of get.py is regression test for this issue.
This is an export from <a href="https://github.com/web-platform-tests/wpt/pull/46584.">https://github.com/web-platform-tests/wpt/pull/46584.</a>

* Source/WebDriver/socket/HTTPServerSocket.cpp:
(WebDriver::HTTPRequestHandler::packHTTPMessage const):
* WebDriverTests/imported/w3c/webdriver/tests/classic/get_title/get.py:
(test_strip_and_collapse):
(test_title_included_entity_references):
(test_title_included_multibyte_char):

Canonical link: <a href="https://commits.webkit.org/279767@main">https://commits.webkit.org/279767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29c4dd444ed4e7d76ea481df9508ba7904bd3e23

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54318 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33721 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6874 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57598 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5050 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5145 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43987 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3362 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56412 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31909 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47042 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25122 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28716 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4364 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3193 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50475 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4573 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59188 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29535 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4808 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51411 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30697 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47135 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50768 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11887 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31670 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30482 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebDriver/socket/HTTPServerSocket.cpp:
```diff
@@ -108,7 +108,13 @@ void HTTPRequestHandler::sendResponse(HTTPRequestHandler::Response&& response)
+        endpoint.send(m_client.value(), reinterpret_cast<const uint8_t*>(span.data()), span.size());
+        return true;
+    });
+    RELEASE_ASSERT(result);
```
If we know in advance that getting UTF-8 is going to be successful, is there some reason to not just use `packet.utf8()`?

Secondly but perhaps more importantly, is this the right layer for making this fix? I am not familiar with this code, but I'm surprised that `RemoteInspectorSocketEndpoint::send` would take a `uint8_t` pointer and a size and then try to interpret that as latin-1. That feels like it's potentially a bug in `RemoteInspectorSocketEndpoint`, not here.

haruhisa-shin replied:
Thank you for the review.

WTFString created by `String::fromUTF8()` keeps the internal data in `m_data16`, not `m_data8`.
In this case, the `.utf8()` will not return the correct result.
The reason for using `tryGetUTF8()` is to look at the result of `is8Bit()` and return the correct result for both `m_data8` and `m_data16`.

There may be some misunderstanding about the latter part.
`RemoteInspectorSocketEndpoint` is, as the name implies, the endpoint of a socket, so it is not responsible for its contents.
The side receiving the data via the socket, in this case the WebDriver user, receives the data as UTF-8 HTML.
Therefore, I don't see any particular problem in sending `uint8_t*` data obtained by `tryGetUTF8()`.

haruhisa-shin replied:
I'm sorry, I made a mistake in my previous comment. 
Your review made me aware of it.

The data given to send was fine with `utf8()`.
I checked `utf8()` and `tryGetUTF8()` again and both returned the same data.
It was the length that was wrong.
I should have used `packet.utf8().length()` indices the length of the encoded string, instead of `packet.length()` indices the length of the original string.

aproskuryakov commented on Source/WebDriver/socket/HTTPServerSocket.cpp:
```diff
@@ -108,7 +108,7 @@ void HTTPRequestHandler::sendResponse(HTTPRequestHandler::Response&& response)
     auto& endpoint = RemoteInspectorSocketEndpoint::singleton();
     auto packet = packHTTPMessage(WTFMove(response));
-    endpoint.send(m_client.value(), packet.utf8().data(), packet.length());
+    endpoint.send(m_client.value(), reinterpret_cast<const uint8_t*>(packet.utf8().data()), packet.utf8().length());
```
This looks like it would do the conversion twice, I suggest putting the `CString` in a temporary variable first.

Otherwise, the code change looks right to me.

haruhisa-shin replied:
You are certainly correct. Thank you for pointing that out.
I fixed it with the commit https://github.com/WebKit/WebKit/commit/637d807860f8ed871baa650ecdaec2e34d289036.

haruhisa-shin committed:
[WebDriver][socket] Titles containing multibyte characters cannot be retrieved correctly
https://bugs.webkit.org/show_bug.cgi?id=270063

Reviewed by Alexey Proskuryakov.

If the document title contains multibyte characters such as Japanese
or entity references, the "Get Title" result will be garbled.

The title string is obtained by JavaScript's "document.title()",
and this data is encoded in UTF8.
However, the StringBuilder.append() function used to create HTTP messages
uses fromLatin1() internally to generate strings from byte data.
This seems to be causing the multibyte characters to be garbled.

This patch changes to use String::fromUTF8() before concatenation to
restore the correct WTF::String even if it contains multibyte characters.

Also, the change of get.py is regression test for this issue.
This is an export from https://github.com/web-platform-tests/wpt/pull/46584.

* Source/WebDriver/socket/HTTPServerSocket.cpp:
(WebDriver::HTTPRequestHandler::packHTTPMessage const):
* WebDriverTests/imported/w3c/webdriver/tests/classic/get_title/get.py:
(test_strip_and_collapse):
(test_title_included_entity_references):
(test_title_included_multibyte_char):

Canonical link: https://commits.webkit.org/279767@main
